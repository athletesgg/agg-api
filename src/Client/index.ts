import feathers from '@feathersjs/feathers'
import authentication, {
  Storage,
  defaultStorage,
} from '@feathersjs/authentication-client'
import socketio from '@feathersjs/socketio-client'
import io from 'socket.io-client'

interface ClientOptions {
  storage?: Storage
  timeout?: number
  url: string
}

class Client implements ClientOptions {
  static DEFAULT_TIMEOUT = 3000

  readonly app: feathers.Application
  readonly socket: SocketIOClient.Socket
  private _timeout: number
  private _url: string

  constructor(options: ClientOptions) {
    const {
      storage = defaultStorage,
      timeout = Client.DEFAULT_TIMEOUT,
      url,
    } : ClientOptions = options

    this._url = url
    this.socket = io(this._url)
    this._timeout = timeout

    this.app = feathers()
    this.app
      .configure(socketio(this.socket, {
        timeout: this._timeout,
      }))
      .configure(authentication({
        path: '/authentication',
        storage,
      }))
  }

  get timeout(): number {
    return this._timeout
  }

  set timeout(timeout: number) {
    this.socket.io.timeout(timeout)

    this._timeout = timeout
  }

  get url(): string {
    return this._url
  }
}

export default Client
