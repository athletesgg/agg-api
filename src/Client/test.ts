import Client from '.'

describe('Client', () => {
  it('should set a URL', () => {
    const url = 'https://api-v3.athletes.gg'

    const client = new Client({
      url,
    })
    expect(client.url).toEqual(url)
  })
})
