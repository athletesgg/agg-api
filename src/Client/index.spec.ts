import Client from '.'
import {
  defaultStorage,
} from '@feathersjs/authentication-client'

jest.mock('socket.io-client', () => (): SocketIOClient.Socket => {
  /* eslint-disable */
  const MockedSocket = require('socket.io-mock')

  const socket = <SocketIOClient.Socket>(new MockedSocket())
  socket.io = <SocketIOClient.Manager>{
    timeout(timeout?: number): SocketIOClient.Manager | number {
      if (!timeout) {
        return 1989
      }

      return this
    }
  }

  return socket
  /* eslint-enable */
})

describe('Client', () => {
  it('should initialize a basic client', () => {
    const url = 'https://api-v3.athletes.gg'

    const client = new Client({
      url,
    })

    expect(client.url).toEqual(url)
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    expect(client.app.get('storage').storage).toEqual(defaultStorage)
    expect(client.timeout).toEqual(Client.DEFAULT_TIMEOUT)
    expect(client).toMatchSnapshot()
  })

  it('should change sio timeout', () => {
    const timeout = 1989

    const client = new Client({
      url: 'https://taylorswift.com/api',
    })

    expect(client.timeout).toEqual(Client.DEFAULT_TIMEOUT)

    client.timeout = timeout
    expect(client.timeout).toEqual(timeout)
  })

  it('should fail to login (no cached credentials)', async () => {
    const client = new Client({
      url: 'https://taylorswift.com/api',
    })

    await expect(client.login()).rejects.toThrowErrorMatchingSnapshot()
  })

  it('should fail to login (timeout)', async () => {
    const client = new Client({
      url: 'https://taylorswift.com/api',
    })

    await expect(client.login({
      password: 'tayloriscool13',
      username: 'betty',
    })).rejects.toThrowErrorMatchingSnapshot()
  })

  it('should fail to logout (timeout)', async () => {
    const client = new Client({
      url: 'https://taylorswift.com/api',
    })

    await expect(client.logout()).rejects.toThrowErrorMatchingSnapshot()
  })
})
