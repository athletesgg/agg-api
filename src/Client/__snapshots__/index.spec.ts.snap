// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client should initialize a basic client 1`] = `
Client {
  "_timeout": 3000,
  "_url": "https://api-v3.athletes.gg",
  "app": Object {
    "_events": undefined,
    "_eventsCount": 0,
    "_maxListeners": undefined,
    "_setup": false,
    "addListener": [Function],
    "authenticate": [Function],
    "authentication": AuthenticationClient {
      "app": [Circular],
      "authenticated": false,
      "options": Object {
        "Authentication": [Function],
        "header": "Authorization",
        "jwtStrategy": "jwt",
        "locationErrorKey": "error",
        "locationKey": "access_token",
        "path": "/authentication",
        "scheme": "Bearer",
        "storage": MemoryStorage {
          "store": Object {},
        },
        "storageKey": "feathers-jwt",
      },
    },
    "configure": [Function],
    "defaultService": [Function],
    "disable": [Function],
    "disabled": [Function],
    "emit": [Function],
    "enable": [Function],
    "enabled": [Function],
    "eventMappings": Object {
      "create": "created",
      "patch": "patched",
      "remove": "removed",
      "update": "updated",
    },
    "eventNames": [Function],
    "get": [Function],
    "getMaxListeners": [Function],
    "hookTypes": Array [
      "before",
      "after",
      "error",
      "finally",
    ],
    "hooks": [Function],
    "init": [Function],
    "io": e {
      "_callbacks": Object {
        "$disconnect": Array [
          [Function],
        ],
      },
      "_emitFn": [Function],
      "broadcast": Object {
        "to": [Function],
      },
      "generalCallbacks": Object {},
      "io": Object {
        "timeout": [Function],
      },
      "joinedRooms": Array [],
      "rooms": Array [],
      "socketClient": e {
        "_emitFn": [Function],
        "_socketMock": [Circular],
      },
    },
    "listenerCount": [Function],
    "listeners": [Function],
    "logout": [Function],
    "methods": Array [
      "find",
      "get",
      "create",
      "update",
      "patch",
      "remove",
    ],
    "mixins": Array [
      [Function],
      [Function],
    ],
    "off": [Function],
    "on": [Function],
    "once": [Function],
    "prependListener": [Function],
    "prependOnceListener": [Function],
    "providers": Array [],
    "rawListeners": [Function],
    "reAuthenticate": [Function],
    "removeAllListeners": [Function],
    "removeListener": [Function],
    "service": [Function],
    "services": Object {},
    "set": [Function],
    "setMaxListeners": [Function],
    "settings": Object {
      "storage": StorageWrapper {
        "storage": MemoryStorage {
          "store": Object {},
        },
      },
    },
    "setup": [Function],
    "use": [Function],
    "version": "4.5.3",
  },
  "socket": e {
    "_callbacks": Object {
      "$disconnect": Array [
        [Function],
      ],
    },
    "_emitFn": [Function],
    "broadcast": Object {
      "to": [Function],
    },
    "generalCallbacks": Object {},
    "io": Object {
      "timeout": [Function],
    },
    "joinedRooms": Array [],
    "rooms": Array [],
    "socketClient": e {
      "_emitFn": [Function],
      "_socketMock": [Circular],
    },
  },
}
`;
